### ISSUE TITLE
Error: Uninitialized string offset 0 \vendor\phpseclib\phpseclib\phpseclib\Net\SSH2.php on line 1110

### ISSUE THREAD

**Issue by smtlab** [phpseclib/phpseclib#1046]
Getting the following error .Please help!

`PHP Notice:  Uninitialized string offset: 0 in \vendor\phpseclib\phpseclib\phpseclib\Net\SSH2.php on line 1110`

`Notice: Uninitialized string offset: 0 in E:\xampp\htdocs\symfony\vendor\phpseclib\phpseclib\phpseclib\Net\SSH2.php on line 1110
PHP Notice:  Expected SSH_MSG_KEXINIT in \vendor\phpseclib\phpseclib\phpseclib\Net\SSH2.php on line 1111`

`Notice: Expected SSH_MSG_KEXINIT in \vendor\phpseclib\phpseclib\phpseclib\Net\SSH2.php on line 1111`

After diggin into the code found that `_get_binary_packet()` function called on line 1104 in SSH2.php returns empty string which is causing this error.
Please help what is the cause of this error.

**Comment by terrafrost** [comment]
Quoting [RFC4253](https://tools.ietf.org/html/rfc4253#page-5), "Key exchange will begin immediately after sending [the SSH-2.0-xxx identification string]". Later, in the same RFC ([Section 7.1](https://tools.ietf.org/html/rfc4253#section-7.1)) it says "Key exchange begins by each side sending the following packet".

If your SSH server is not sending the key exchange packet immediately than I'd say your server is buggy. I would actually be curious to know what your server identification string is. You can get it by doing `echo $ssh->getServerIdentification();`

That said, it is a fact that some SSH servers are buggy and sometimes workarounds can be implemented to work on those servers. PuTTY has a bunch of workarounds present:

https://the.earth.li/~sgtatham/putty/0.60/htmldoc/Chapter4.html#config-ssh-bugs

phpseclib doesn't need all of these and, in some cases, needs workarounds that PuTTY doesn't, because phpseclib supports ciphers that PuTTY doesn't.

Long story short, try this:

https://github.com/terrafrost/phpseclib/tree/2.0-send-kex-before-rcpt

**Comment by smt92** [comment]
Thank you for replying.

I tested with multiple servers getting same error. Also I can connect with PuTTY with no issues.
Server identification Id: SSH-2.0-OpenSSH_5.3

**Comment by terrafrost** [comment]
Did you try the branch I made?:

https://github.com/terrafrost/phpseclib/tree/2.0-send-kex-before-rcpt

**Comment by foxymoron-repos** [comment]
I think problem is with my home network / ISP. My ISP block all outgoing traffic.
I tested with my office network it works.

**Comment by terrafrost** [comment]
Ah. Well, FWIW, I did forget to actually commit the change in the branch I made lol. I just committed / pushed but yah if it's your home network / ISP than that branch wouldn't help.

**Comment by bantu** [comment]
This is probably a bug that should be fixed nonetheless.

**Comment by terrafrost** [comment]
I wouldn't call it a bug. It's an alternative way of doing things that shouldn't be necessary for properly implemented SSH servers. And thus far I am aware of no instances where this would actually make a difference. I hypothesized that it might help here but if the OP isn't having issues in his office network then idk.

Like I'll merge it if it actually helps but if it doesn't then it's just a change for the sake of making a change lol

**Comment by bantu** [comment]
@terrafrost What I was trying to say is that if $s can be empty, you should not access $s[0].

**Comment by smtlab** [comment]
I agree atleast proper error message or exception handling seems valid here. Since its very difficult to track down whats going wrong just by general php errors.

**Comment by terrafrost** [comment]
> @terrafrost What I was trying to say is that if $s can be empty, you should not access $s[0].

If we're going to do that we might as well fix all the PHP warnings that'll be emitted if 

https://github.com/terrafrost/phpseclib/tree/1.0-suppress-unpack-errors

Demo showing that unpack on an insufficiently long string will result in a warning:

```php
function string_shift(&$string, $index = 1)
{
    $substr = substr($string, 0, $index);
    $string = substr($string, $index);
    return $substr;
}

$str = '';
unpack('N', string_shift($str, 4));
```

**Comment by solazs** [comment]
@terrafrost We've tried your branch (https://github.com/terrafrost/phpseclib/tree/2.0-send-kex-before-rcpt) and it did fix the said problem above. Our script connects to some cisco switches, so fixing the "buggy" ssh server is not an option for us.

+1 one for getting this merged.

**Comment by solazs** [comment]
More info: this bug shows up with version 2.0.5, no such problem with 2.0.4.

**Comment by mikebrandl** [comment]
@terrafrost I've also tried your branch (https://github.com/terrafrost/phpseclib/tree/2.0-send-kex-before-rcpt) and it did fix the said problem. Our script connects to some cisco routers, so fixing the "buggy" ssh server is not an option for us.

+1 one for getting this merged.

Also this is still present in 2.0.6, and connects properly using the SSH File from your branch. 

I do get the error connection closed prematurely by doing it this way.

**Comment by terrafrost** [comment]
I'll try to get https://github.com/terrafrost/phpseclib/commit/8458c94f50ed3a75a2154911a1bb1c47a6c61635 merged in this weekend. That said, I'd like to hold off on doing a release for a while to give people a chance to test this change.

**Comment by terrafrost** [comment]
I've created a new branch that should fix this: #1162

**Comment by mikebrandl** [comment]
Awesome Thanks @terrafrost When do you think this might get merged?

**Comment by terrafrost** [comment]
@mikebrandl - probably this evening. I need to merge this into 2.0 and master. I posted my comment at 12:31am, local time, and didn't want to stay up any later than I had already since I had to be up ~8h later to get to work on time.

**Comment by mikebrandl** [comment]
Thanks @terrafrost Much Appreciated!

**Comment by terrafrost** [comment]
#1162 has been merged.
